org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/Conversions.java

```diff
@@ -61,139 +189,305 @@ public static Object doWrapArray(Object object) {
+		if(value==null)return null;
+		return value.internalToArray();
+	}
+
+	/**
+	 * Unwraps {@code object} to extract the original array if and only if
+	 * {@code object} was previously created by {@link #doWrapArray(Object)}. If
+	 * the array's component type cannot be determined at runtime,
+	 * {@link Object} is used.
+	 * 
+	 * @param value the object to be unwrapped. May be <code>null</code>.
+	 * @return the previously wrapped array if the given value represents such.
+	 * Otherwise returns the value unmodified. May return <code>null</code> if
+	 * the value was <code>null</code>.
+	 */
+	@Pure
+	public static float[] unwrapArray(WrappedFloatArray value) {
+		if(value==null)return null;
+		return value.internalToArray();
+	}
+
+	/**
+	 * Unwraps {@code object} to extract the original array if and only if
+	 * {@code object} was previously created by {@link #doWrapArray(Object)}. If
+	 * the array's component type cannot be determined at runtime,
+	 * {@link Object} is used.
+	 * 
+	 * @param value the object to be unwrapped. May be <code>null</code>.
+	 * @return the previously wrapped array if the given value represents such.
+	 * Otherwise returns the value unmodified. May return <code>null</code> if
+	 * the value was <code>null</code>.
 	 */
 	@Pure
 	public static Object unwrapArray(Object value) {
+		if(value==null)return null;
 		return unwrapArray(value, Object.class);
 	}
 
 	/**
-	 * Unwraps {@code object} to extract the original array if and only if {@code object} was previously created by
-	 * {@link #doWrapArray(Object)}.
+	 * Unwraps {@code object} to extract the original array if and only if
+	 * {@code object} was previously created by {@link #doWrapArray(Object)}.
 	 * 
-	 * @param value
-	 *            the object to be unwrapped. May be <code>null</code>.
-	 * @param componentType
-	 *            the expected component type of the array. May not be <code>null</code>.
-	 * @return the previously wrapped array if the given value represents such. Otherwise returns the value unmodified.
-	 *         May return <code>null</code> if the value was <code>null</code>.
-	 * @throws ArrayStoreException
-	 *             if the expected runtime {@code componentType} does not match the actual runtime component type.
+	 * @param value the object to be unwrapped. May be <code>null</code>.
+	 * @param componentType the expected component type of the array. May not be
+	 * <code>null</code>.
+	 * @return the previously wrapped array if the given value represents such.
+	 * Otherwise returns the value unmodified. May return <code>null</code> if
+	 * the value was <code>null</code>.
+	 * @throws ArrayStoreException if the expected runtime {@code componentType}
+	 * does not match the actual runtime component type.
 	 */
 	@Pure
-	public static Object unwrapArray(Object value, Class<?> componentType) {
+	public static <T,K extends T> Object unwrapArray(Iterable<K> value, Class<T> componentType) {
```

Comment:
I don't think the signature is correct.
Assuming you have something like an array of strings and you store it in an iterable of objects.
Now K would point to object and T would need to be String. That does not fulfil the type constraints.
Also I don't see the parameter componentType to be used anymore, e.g. checkComponentType is missing.

Reply:
so the orginal method that ignore componentType is wrong
`
if (value instanceof WrappedArray<?>) {
			Object result = ((WrappedArray<?>) value).internalToArray();
			return checkComponentType(result, componentType);
		}

`
and the check
`	private static Object checkComponentType(Object array, Class<?> expectedComponentType) {
		Class<?> actualComponentType = array.getClass().getComponentType();
		if (!expectedComponentType.isAssignableFrom(actualComponentType)) {
			throw new ArrayStoreException(
					String.format("The expected component type %s is not assignable from the actual type %s",
							expectedComponentType.getCanonicalName(), actualComponentType.getCanonicalName()));
		}
		return array;
	}`

Reply:
this xtext test fail also with master code
 ```
@Test def fail(){
        var int[] array = #[0,1]
        var List<Integer> list = array
        var Integer[] array2 = list
        array2.forEach[v, index|
            assertEquals(index,v)
        ]        
    }
```

Reply:
Does it fail with an assertion error or with a ClassCastException?

Reply:
What do you mean with ‚original method that ignore checkComponentType‘?

Reply:
the method on master if input value is WrappedXXXArray don't convert result to destination type
like it does for other Iterable

Reply:
Sorry, I can’t follow. I see invocations of checkComponentType on master at the (to me) relevant locations.